addons/l10n_id_efaktur/models/efaktur.py

```diff
@@ -0,0 +1,109 @@
+
+    @api.model
+    def pop_number(self, company_id):
+        range = self.search([('company_id', '=', company_id)], order="min ASC", limit=1)
+        if not range:
+            return None
+
+        popped = int(range.min)
+        if int(range.min) >= int(range.max):
+            range.unlink()
+        else:
+            range.min = '%013d' % (popped + 1)
+        return popped
+
+    @api.model
+    def push_number(self, company_id, number):
+        return self.push_numbers(company_id, number, number)
+
+    @api.model
+    def push_numbers(self, company_id, min, max):
+        range_sup = self.search([('min', '=', '%013d' % (int(max) + 1))])
+        if range_sup:
+            range_sup.min = '%013d' % int(min)
+            max = range_sup.max
+
+        range_low = self.search([('max', '=', '%013d' % (int(max) - 1))])
+        if range_low:
+            range_sup.unlink()
+            range_low.max = '%013d' % int(max)
+
+        if not range_sup and not range_low:
+            self.create({
+                'company_id': company_id,
+                'max': '%013d' % int(max),
+                'min': '%013d' % int(min),
+            })
+
+
+    @api.constrains('min', 'max')
+    def _constrains_min_max(self):
+        for record in self:
+            if not len(record.min) == 13 or not len(record.max) == 13:
+                raise ValidationError(_("There should be 13 digits in each number."))
+
+            if record.min[:-8] != record.max[:-8]:
+                raise ValidationError(_("First 5 digits should be same in Start Number and End Number."))
+
+            if int(record.min[-8:]) > int(record.max[-8:]):
+                raise ValidationError(_("Last 8 digits of End Number should be greater than the last 8 digit of Start Number"))
+
+            if (int(record.max) - int(record.min)) > 10000:
+                raise ValidationError(_("The difference between the two numbers must not be greater than 10.000"))
+
+            # The number of records should always be very small, so it is ok to search in loop
+            if self.search([
+                '&', ('id', '!=', record.id), '|', '|',
+                '&', ('min', '<=', record.max), ('max', '>=', record.max),
+                '&', ('min', '<=', record.min), ('max', '>=', record.min),
+                '&', ('min', '>=', record.min), ('max', '<=', record.max),
+            ]):
+                raise ValidationError(_('Efaktur interleaving range detected'))
+
+    @api.depends('min', 'max')
+    def _compute_available(self):
+        for record in self:
+            record.available = 1 + int(record.max) - int(record.min)
+
+    @api.depends('company_id')
+    def _compute_default(self):
+        for record in self:
+            query = """
+                SELECT MAX(SUBSTRING(l10n_id_tax_number FROM 4))
+                FROM account_move
+                WHERE l10n_id_tax_number IS NOT NULL
+                  AND company_id = %s
+            """
```

Comment:
Is not this a bit overkill?  A search on the existing efaktur ranges might be enough.  Or am I missing something?